integrations/clickup/clickup/client.py

```diff
@@ -0,0 +1,124 @@
+        logger.info("Getting teams from Clickup")
+        url = f"{self.clickup_url}/team"
+        team_response = await self.client.get(url)
+        team_response.raise_for_status()
+        team_list = team_response.json()["teams"]
+        event.attributes[CacheKeys.TEAMS] = team_list
+        yield team_list
+
+    async def get_spaces(self) -> AsyncGenerator[list[dict[str, Any]], None]:
+        logger.info("Getting spaces from Clickup")
+        team_list: list[dict[str, Any]] = event.attributes.get(CacheKeys.TEAMS, [])
+        if not team_list:
+            async for team_list in self.get_teams():
+                async for spaces in self._fetch_spaces_for_teams(team_list):
+                    yield spaces
+        else:
+            async for spaces in self._fetch_spaces_for_teams(team_list):
+                yield spaces
+    
+    async def _fetch_spaces_for_teams(self, team_list: list[dict[str, Any]]) -> AsyncGenerator[list[dict[str, Any]], None]:
+        for team_dict in team_list:
+            url = f'{self.clickup_url}/team/{team_dict["id"]}/space'
+            space_response = await self.client.get(url)
+            space_response.raise_for_status()
+            space_list = space_response.json()["spaces"]
+            event.attributes[CacheKeys.TEAM] = team_dict
+            yield space_list
+    
+    async def get_folders(self) -> AsyncGenerator[list[dict[str, Any]], None]:
+        logger.info("Getting folders from Clickup")
+        async for space_list in self.get_spaces():
+            for space_dict in space_list:
+                url = f'{self.clickup_url}/space/{space_dict["id"]}/folder'
+                folder_response = await self.client.get(url)
+                folder_response.raise_for_status()
+                folder_list = folder_response.json()["folders"]
+                yield folder_list
+
+    async def get_projects(self) -> AsyncGenerator[list[dict[str, Any]], None]:
+        logger.info("Getting projects from Clickup")
+        async for folder_list in self.get_folders():
+            async for projects in self._fetch_projects_for_folders(folder_list):
+                yield projects
+        async for space_list in self.get_spaces():
+            async for projects in self._fetch_projects_for_spaces(space_list):
+                yield projects
+    
+    async def _fetch_projects_for_folders(self, folder_list: list[dict[str, Any]]) -> AsyncGenerator[list[dict[str, Any]], None]:
+        for folder_dict in folder_list:
+            url = f'{self.clickup_url}/folder/{folder_dict["id"]}/list'
+            project_response = await self.client.get(url)
+            project_response.raise_for_status()
+            project_list = project_response.json()["lists"]
+            team: dict[str, Any] = event.attributes.get(CacheKeys.TEAM, {})
+            project_list_with_team = [{**project, "__team": team} for project in project_list]
+            self._cache_projects(project_list_with_team)
```

Comment:
For caching functionality, ocean provides a decorator to make this easy. See how we've done it [here](https://github.com/port-labs/ocean/blob/f032aa8b77c1a15cdecfa866719832d7fb167761/integrations/sentry/clients/sentry.py#L89)